### PR TITLE
feat: Prompt Builder as campaign child route + shared context helper (fixes #212)

### LINKED ISSUE #184: Feature: Prompt Builder - assemble context-rich AI prompts from campaign and party data
## Overview

The [dm-dashboard](https://github.com/aikithoughts/dm-dashboard) includes a **Prompt Builder** — a structured form that assembles context-rich prompts the DM can paste into an AI assistant (e.g. Claude). It automatically incorporates the active campaign name, module, current chapter, and party composition so the DM doesn't have to re-type that context every time.

Session-combat has rich party, character, monster, and encounter data that is ideal context for such prompts. Adding a Prompt Builder would let DMs generate NPCs, location descriptions, magic items, shop write-ups, and room descriptions directly from their session data.

## Goals

- DM can select a prompt type and fill in a short form
- App assembles a complete, context-rich prompt using live campaign and party data
- DM can copy the prompt to the clipboard in one click
- Generated prompts can optionally be saved to a Content Library (see related issue)

## Initial Prompt Templates to Support

| Template | Key fields |
|---|---|
| NPC | role/occupation, location, special requirements |
| Location Description | type, atmosphere, details to include |
| Shop / Establishment | type, setting, special requirements |
| Magic Item | item type, rarity, thematic flavor, special requirements |
| Room Description | room name/number, original purpose, physical details, inhabitants |

Each template injects the active campaign name, module, current chapter, and party member list into the prompt automatically.

### Step 1: Prerequisites
- Campaign Management (#182) must be completed so the builder can read the active campaign and chapter
- The active campaign and party data should be fetchable via existing APIs

### Step 2: Define prompt templates
- Create `lib/prompts/templates.ts` containing the five initial template definitions
- Each template exports: `label`, `fields[]` (key, label, placeholder, optional flag), and a `build(fields, campaign, party)` function that returns the assembled prompt string

### Step 3: API (optional — no backend required for generation)
- Prompt assembly happens entirely client-side using data already fetched
- The only API needed is saving to the Content Library (covered in the Content Library issue)

### Step 4: Prompt Builder UI
- Add `app/prompts/page.tsx`
- Template selector tabs at the top (NPC / Location / Shop / Magic Item / Room)
- Dynamic form below based on selected template
- "Generate Prompt" button assembles and displays the prompt in a read-only textarea
- "Copy to Clipboard" button
- "Save to Library" button (wired up once Content Library issue is implemented)
- Add a **Prompt Builder** nav link in `app/layout.tsx`

### Step 5: Tests
- Unit tests for each `build()` function in `lib/prompts/templates.ts` — verify all expected fields are present in the output
- Integration test: render the page, select each template, fill form, click Generate, assert prompt contains campaign name and party members
- Test that clipboard copy works (or mock navigator.clipboard)

## Reference

Proposed UI integration spot

Image

dm-dashboard implementation:
- Templates + builder: `app/prompts/page.js` — complete implementation with all five templates
- Each template defines `fields[]` and a `build(fields, campaign, party)` function
- The NPC, Location, Shop, Magic Item, and Room templates are ready to port to TypeScript

### LINKED ISSUE #212: Bug: Session logs only uses first linked party — silently ignores others when multiple parties share a campaign
## Bug

`app/campaigns/[id]/sessions/page.tsx` uses `Array.find()` to locate the party linked to the current campaign:

```typescript
// line 330
const linked = partiesData.find(p => p.campaignId === campaignId) ?? null;
setParty(linked);
```

`find()` returns the **first match only**. If a DM runs multiple parties through the same campaign (e.g. two separate play groups in the same adventure), only one party's members are used. The rest are silently dropped.

## Impact

- **NPC event auto-population** (`buildNpcEventsFromMemberChanges`) is called with an incomplete member list — only the first party's join/leave events are suggested when creating a new session log entry
- The "No linked party found" warning (line 171) never appears even if the *relevant* party isn't the first one returned, giving false confidence that party context is loaded
- Session log entries end up missing member-change events for all but the first party

## Expected behaviour

All parties linked to the campaign (`p.campaignId === campaignId`) should be found and their member lists merged before being passed to `buildNpcEventsFromMemberChanges`.

## Fix direction

Replace `find()` with `filter()` and aggregate members across all linked parties:

```typescript
const linkedParties = partiesData.filter(p => p.campaignId === campaignId);
// merge members from all parties before passing downstream
```

## Note

This fix should be implemented via a **shared campaign context helper** (`lib/utils/campaignContext.ts` or similar) that returns `{ campaign, chapter, parties, characters }` for a given `campaignId`. This helper will also be used by the Prompt Builder (#184) — see that issue for the full design. Fixing this bug in isolation before the helper exists risks duplicating logic; the recommended order is:

1. Implement Prompt Builder (#184), which includes the shared helper
2. Refactor session logs to consume the same helper, picking up this fix for free

If this needs to be fixed independently before #184 lands, the fix is straightforward — just replace `find()` with `filter()` and aggregate the member arrays.

## Related

- #184 Prompt Builder (will introduce the shared helper)

### PR DESCRIPTION
**Pull request by dougis**
## Summary

Adds a Prompt Builder page as a campaign child route, a shared `fetchCampaignContext` helper + `useCampaignContext` hook, and five prompt templates (NPC, Location, Shop, Magic Item, Room Description). Also refactors the Sessions page to use the shared context helper, fixing multi-party bug #212.

Closes #184
Fixes #212

### New: Shared Context Helper
- `lib/utils/campaignContext.ts` — `fetchCampaignContext(campaignId)` parallel-fetches campaign, parties, and characters; merges members from **all** linked parties (fixes #212); resolves `currentChapterId` → `CampaignChapter`; filters soft-deleted characters
- `lib/hooks/useCampaignContext.ts` — React hook wrapping the above with loading/error state and `refresh()` trigger
- `lib/types.ts` — new `CampaignContext` and `BuiltPrompt` interfaces

### New: Prompt Templates
- `lib/prompts/templates.ts` — `PromptField`, `PromptTemplate` interfaces; `buildSystemPrompt()` assembles campaign/chapter/party block; 5 named templates each with `fields[]` + `build(fields, context) → BuiltPrompt`; agentic-ready `{ systemPrompt, userMessage, fullText }` output shape

### New: Prompt Builder Page
- `app/campaigns/[id]/prompts/page.tsx` — template tabs, dynamic field form, Generate → read-only textarea, Copy to clipboard with "Copied!" feedback, Save to Library stub (disabled, references issue #185), loading/error/no-party states

### Changed: Campaigns List Page
- `app/campaigns/page.tsx` — adds "Prompt Builder" link alongside "Sessions" in each campaign card

### Refactored: Sessions Page (fixes #212)
- `app/campaigns/[id]/sessions/page.tsx` — replaces manual `Array.find()` party lookup with `useCampaignContext`; `buildNpcEventsFromMemberChanges` now receives `allMembers` merged from **all** linked parties

## Tests

40 new test cases across 6 new/updated test files:

| File | Tests | Coverage |
|------|-------|----------|
| `tests/unit/utils/campaignContext.test.ts` | 9 | multi-party merge, soft-delete filter, parallel fetch, chapter resolution, error rejection |
| `tests/unit/hooks/useCampaignContext.test.ts` | 4 | loading state, success, error, refresh |
| `tests/unit/prompts/templates.test.ts` | 23 | all 5 templates, buildSystemPrompt, edge cases |
| `tests/integration/prompts/promptBuilder.test.tsx` | 12 | C1-1 through C1-12: full page render, tabs, generate, copy, save stub, loading/error/no-party |
| `tests/integration/sessions/sessionLogs.test.tsx` | 5 | D1 regression (single-party) + D2 multi-party |
| `tests/unit/components/SessionsPage.test.tsx` | updated | adapted for hook-based architecture |

**Validation:** 1424 unit tests pass · 17 integration tests pass · TypeScript clean · build succeeds · 0 lint errors

### CODE REVIEW

File: tests/integration/prompts/promptBuilder.test.tsx [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+    // Don't fill fields - click Generate immediately
+    const generateBtn = Array.from(container.querySelectorAll('button')).find(b => b.textContent?.includes('Generate'));
+    await act(async () => { generateBtn?.click(); });
+
+    // The mock template has no required fields (optional field only), so this test checks our validation logic
+    // Note: if mock template has no required fields, no validation fires — that's expected
+    // The page should handle gracefully
+    expect(container.textContent).toBeDefined();
+  });
```
[the remark this answers is not in the document]
Comment (dougis) [reply tests/integration/prompts/promptBuilder.test.tsx:165]: Fixed — the `jest.mock('@/lib/prompts/templates', ...)` call is now at the top of the file (module-level, not inside the test). The C1-5 test properly switches to the Location Description tab (which has a required `locationName` field with no `optional: true`), clicks Generate without filling it, and asserts both that no read-only textarea appears and that the error message contains 'Location Name'. The validation path is exercised correctly.

File: app/campaigns/[id]/sessions/page.tsx [hunk 2]
```diff
@@ -112,9 +115,9 @@ function SessionForm({
   const [error, setError] = useState<string | null>(null);
 
   useEffect(() => {
-    if (existing || !party) return;
-    setEvents(buildNpcEventsFromMemberChanges(party.members, lastSessionDate));
-  }, [party, lastSessionDate, existing]);
+    if (existing || !hasParty) return;
+    setEvents(buildNpcEventsFromMemberChanges(allMembers, lastSessionDate));
```
[the remark this answers is not in the document]
Comment (dougis) [reply app/campaigns/[id]/sessions/page.tsx:124]: Covered by test D2-1 in `tests/integration/sessions/sessionLogs.test.tsx` — it renders the page with two parties (PARTY_ALICE_BOB and PARTY_CAROL), opens a new session, and asserts that `buildNpcEventsFromMemberChanges` is called with all three members combined from both parties (`c-1`, `c-2`, `c-3`). This directly validates the fix for #212.

File: lib/hooks/useCampaignContext.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      setError(err instanceof Error ? err.message : 'Failed to load campaign context');
+      setContext(null);
+    } finally {
+      setLoading(false);
+    }
+  }, [campaignId]);
+
+  useEffect(() => {
+    void load();
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
[the remark this answers is not in the document]
Comment (dougis) [reply lib/hooks/useCampaignContext.ts]: The eslint-disable comment was removed in a prior commit. The effect at line 34–36 depends only on `load` and `tick`, both of which are correctly listed in the dependency array — no disable needed.

File: tests/unit/utils/campaignContext.test.ts [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+      else if (s.includes('/api/characters')) callOrder.push('characters');
+      await Promise.resolve();
+      return { ok: true, json: async () => s.includes('/api/campaigns') ? campaign : [] } as unknown as Response;
+    }) as typeof fetch;
+
+    await fetchCampaignContext('camp-1', parallelFetch);
+
+    expect(parallelFetch).toHaveBeenCalledTimes(3);
+    // All three should have been registered before any awaited result consumed
+    expect(callOrder).toHaveLength(3);
```
[the remark this answers is not in the document]
Comment (dougis) [reply tests/unit/utils/campaignContext.test.ts]: Fixed — A2-8 now uses deferred promises. The fetch mock returns unresolved `Promise<Response>` objects and records each call in `started[]`. After yielding to the microtask queue, we assert all three entries are present before resolving any of them. A sequential implementation would stall at the first `await`, never starting the other two, making the assertions on `started` fail.

File: tests/integration/prompts/promptBuilder.test.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (dougis) [reply tests/integration/prompts/promptBuilder.test.tsx:165]: Fixed — the mock is module-level only. C1-5 switches to the Location Description tab (required `locationName` field), clicks Generate without filling, then asserts `textarea[readonly]` is null and the error message contains 'Location Name'.

File: tests/unit/hooks/useCampaignContext.test.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
+/**
+ * @jest-environment jsdom
+ */
+(globalThis as unknown as Record<string, unknown>).IS_REACT_ACT_ENVIRONMENT = true;
+
+import React from 'react';
+import { act } from 'react';
+import { createRoot } from 'react-dom/client';
+import { describe, test, expect, jest, beforeEach, afterEach } from '@jest/globals';
```
[the remark this answers is not in the document]
Comment (dougis) [reply tests/unit/hooks/useCampaignContext.test.ts]: The unused `afterEach` import was already removed — the import statement on line 9 now only includes `describe`, `test`, `expect`, `jest`, and `beforeEach`.

File: openspec/changes/prompt-builder-campaign-child/tasks.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+- [ ] All tasks in groups A–D marked complete
+- [ ] All steps in Remote push validation below
+
+## Remote push validation
+
+Verification requirements (all must pass before PR or pushing updates to a PR):
+
+- **Unit tests** — `npm run test` — all tests must pass
+- **Integration tests** — `npm run test:integration` — all tests must pass
+- **Build** — `npm run build` — must succeed with no errors
```
[the remark this answers is not in the document]
Comment (dougis) [reply openspec/changes/prompt-builder-campaign-child/tasks.md:110]: Already corrected — the Validation and Remote push validation sections now reference `npm run test:unit` and `npm run test:integration` specifically.

File: lib/utils/campaignContext.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  const allMembers = parties.flatMap(p => p.members);
+  const memberIds = new Set(allMembers.map(m => m.characterId));
+
+  const chapter = campaign.currentChapterId
+    ? (campaign.chapters.find(c => c.id === campaign.currentChapterId) ?? null)
+    : null;
+
+  const characters = allCharacters.filter(
+    c => memberIds.has(c.id) && !c.deletedAt,
+  );
```
[the remark this answers is not in the document]
Comment (dougis) [reply lib/utils/campaignContext.ts:44]: Implemented — `lib/utils/campaignContext.ts` line 27 filters to active-only members when resolving characters: `const memberIds = new Set(allMembers.filter(m => !m.leftAt).map(m => m.characterId))`. The full `allMembers` array (including departed members) is preserved unchanged for session-event timeline diffing, as noted in the inline comment on line 25–27. Covered by test A2-7b.
